File: 0155-min-stack/0155-min-stack.js

var MinStack = function() {
    this.stack = [];
    this.iStack = [];
};

/** 
 * @param {number} val
 * @return {void}
 */
MinStack.prototype.push = function(val) {
    this.stack.push(val)
    if (this.iStack.length){
        if (val <= this.stack[this.iStack[this.iStack.length - 1]]) {
            this.iStack.push(this.stack.length - 1)
        } else {
            this.iStack.push(this.iStack[this.iStack.length - 1])
        }
    } else {
        this.iStack.push(0)
    }
};

/**
 * @return {void}
 */
MinStack.prototype.pop = function() {
    this.stack.pop();
    this.iStack.pop();
};



/**
 * @return {number}
 */
MinStack.prototype.top = function() {
    return this.stack[this.stack.length - 1]
};

/**
 * @return {number}
 */
MinStack.prototype.getMin = function() {
    return this.stack[this.iStack[this.iStack.length - 1]]
};

/** 
 * Your MinStack object will be instantiated and called as such:
 * var obj = new MinStack()
 * obj.push(val)
 * obj.pop()
 * var param_3 = obj.top()
 * var param_4 = obj.getMin()
 */

File: 0155-min-stack/0155-min-stack.py
class MinStack:

    def __init__(self):
        self.stack = []
        self.iStack = []

    def push(self, val: int) -> None:
        self.stack.append(val)
        val = min(val, self.iStack[-1] if self.iStack else val)
        self.iStack.append(val)

    def pop(self) -> None:
        self.stack.pop()
        self.iStack.pop()

    def top(self) -> int:
        return self.stack[-1]

    def getMin(self) -> int:
        return self.iStack[-1]
        


# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(val)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()